x/zkism/types/ism.go

```diff
@@ -46,8 +51,51 @@ func (ism *ZKExecutionISM) Verify(ctx context.Context, metadata []byte, message
 }
 
 // verifyZKProof verifies a ZK proof to update the ISM's state root and height.
-func (ism *ZKExecutionISM) verifyZKProof(_ ZkExecutionISMMetadata) (bool, error) {
-	// TODO: https://github.com/celestiaorg/celestia-app/issues/4723
+func (ism *ZKExecutionISM) verifyZKProof(metadata ZkExecutionISMMetadata) (bool, error) {
+	groth16VkHash := sha256.Sum256(ism.StateTransitionVerifierKey)
+	if !bytes.Equal(groth16VkHash[:4], metadata.Proof[:4]) {
+		return false, fmt.Errorf("prefix mismatch: first 4 bytes of verifier key hash (%x) do not match proof prefix (%x)", groth16VkHash[:4], metadata.Proof[:4])
+	}
+
+	proof, err := groth16.UnmarshalProof(metadata.Proof[4:])
+	if err != nil {
+		return false, err
+	}
+
+	vk, err := groth16.NewVerifyingKey(ism.StateTransitionVerifierKey)
+	if err != nil {
+		return false, err
+	}
+
+	// TODO: We should store the sp1 verifier key on the ism.
```

Comment:
We need to store both the groth16 verifier key and the sp1 verifier key on the ism, is that correct? cc. @cmwaters

Reply:
Understood, the conversion code is in place here already. But we will need to update the ism protobuf definition to include the additional field